python-api: further api work

mlyle opened the pull request:
The focus of this one was to factor the log handling code into the uavo talk subsystem, and to create a magical module that can be imported to run these scripts standalone instead of only using ipython.

This also separates logview.py into shell.py, dump.py, and logview.py for the different modes.  Note I am unable to test logview.py in my environment.

peabody124 commented:
Note to self: verify ```logview.py -v``` still works.

mlyle commented:
Rebased to next to escape conflicts.

mlyle commented:
Oh, also note, it's just logview.py, shell.py, and dump.py for the 3 different modes.  no -v anymore

BrainFPV commented on python/taulabs/telemetry.py:
```diff
@@ -177,28 +199,62 @@ def serviceConnection(self, timeout=None):
+            return None
 
-        self.__select(0)
+        ret=self.recv_buf
```
style: space around `=`

BrainFPV commented on python/taulabs/telemetry.py:
```diff
@@ -177,28 +199,62 @@ def serviceConnection(self, timeout=None):
 
-        self.__select(0)
+        ret=self.recv_buf
+        self.recv_buf=''
```
same

BrainFPV commented on python/shell.py:
```diff
@@ -0,0 +1,45 @@
+#!/usr/bin/python -B
+
+from taulabs import telemetry,uavo
```
style: space after `,`

BrainFPV commented:
Looks good to me

peabody124 commented on python/logview.py:
```diff
@@ -1,180 +1,22 @@
-            e("Analyzing log file: %s" % src)
+    # Start the log viwer app
+    from PyQt4 import QtGui
+    from logviewer.gui import Window
```
we need to ```import``` sys here

mlyle replied:
Good catch.

peabody124 commented on python/mag_calibration.py:
```diff
@@ -1,3 +1,7 @@
+#!/usr/bin/python
```
how would you feel about leaving mag_calibration in a calibration directory? I can see us having a number of calibration routines and it might be good to organize them. is that a problem for importing?

mlyle replied:
Just adds a line of code to set the path (on each importer).

peabody124 commented on python/taulabs/readbyargs.py:
```diff
@@ -0,0 +1,3 @@
+import telemetry
```
is this method used anywhere or callable? what does this do?

mlyle replied:
So, when you import a python module, the code runs.

This creates a shorthand when you do from taulabs.readbyargs import uavo_list ... that it runs that method.  It's slight abuse of the module system but seemed like a convenient way to do this.

mlyle replied:
(the standalone programs other than minimal_connection do things this way)

peabody124 commented on python/taulabs/telemetry.py:
```diff
@@ -9,35 +9,40 @@
         self.uavo_defs = uavo_defs
         self.uavtalk_generator = uavtalk.processStream(uavo_defs,
-            useWallTime=useWallTime)
+            useWallTime=useWallTime, logTimestamps=weirdTimestamps)
```
name could be more descriptive ;-)

peabody124 commented:
small comments, but essentially good to merge once they are addressed

mlyle commented:
Fixed up according to review feedback, and moved mag_cal script.  Check out the change to that-- if the new wordy boilerplate is OK I'm happy for commit.

(Also, we should be careful to cut and paste that boilerplate exactly, just in case i need to sed it with anything subsequently)

peabody124 commented:
LGTM. @solidgoldbomb you want to review this at all? I think it should be good, though.

solidgoldbomb commented:
I'll look it over later today.

solidgoldbomb commented on python/calibration/mag_calibration.py:
```diff
@@ -97,4 +100,17 @@ def distortion(x,mag_x=mag_x,mag_y=mag_y,mag_z=mag_z,LH=LH,LV=LV):
+	print mag_calibration(uavo_list.as_numpy_array(UAVO_Magnetometer), uavo_list.as_numpy_array(UAVO_Gyros))
+
+	# Wait for user to close window.
+        matplotlib.pyplot.show()
```
indentation (should be tabs)

solidgoldbomb commented on python/calibration/mag_calibration.py:
```diff
@@ -97,4 +100,17 @@ def distortion(x,mag_x=mag_x,mag_y=mag_y,mag_z=mag_z,LH=LH,LV=LV):
+if __name__ == "__main__":
+	import sys, os
+	sys.path.insert(1, os.path.dirname(sys.path[0]))
+        from taulabs.readbyargs import uavo_list
```
indentation (should be tabs)

solidgoldbomb commented on python/calibration/mag_calibration.py:
```diff
@@ -1,100 +1,116 @@
+if __name__ == "__main__":
+        import sys, os
+        sys.path.insert(1, os.path.dirname(sys.path[0]))
+        from taulabs.readbyargs import uavo_list
```
Eww, uavo_list is global and initialized as a side effect of importing this?  We need a better way to deal with this.

mlyle replied:
So the existing stuff that ran in the ipython shell expect uavo_list to be global-- that's how it's been provided to them through the ipython instantiation.  This was an effort to come up with a minimal adapter to allow them to be run either way.

mlyle replied:
(Though in this case those lines could easily/safely be moved to main)

solidgoldbomb replied:
We should set a precedent that all of these helper scripts accept the ```uavo_list``` as an argument.  That way, the shell can still dump uavo_list into the namespace so it's easy to work with interactively but then require the helper script to be called like ```mag_calibration(uavo_list)```.  The ```main``` can then be responsible for constructing ```uavo_list``` and passing it as a parameter to the ```mag_calibration``` function.

solidgoldbomb commented on python/calibration/mag_calibration.py:
```diff
@@ -1,100 +1,116 @@
+        ideally when constantly turning to spread the data around that
+        sphere somewhat evenly (or at least in a horizontal plane)"""
+
+        import numpy
```
is this still required?

mlyle replied:
Unfortunately yes, though some of hte later ones aren't.  The top lines match the conventions in how ipython is invoked, while the code in the function assumes that numpy stuff will be in the numpy namespace.  Both are used (numpy.array, minimize, etc).

I'd be glad to clean this up in a subsequent pull.

Importing * from numpy is kinda bad, because htere's a big polluted namespace in there.  Most numpy code i've seen imports numpy as np to balance the namespace issues vs. the wordiness.

solidgoldbomb replied:
Agreed, we should be doing ```import numpy as np``` and fixing up all references to it.  Ok to defer to later PR of course.

solidgoldbomb commented on python/minimal_connection.py:
```diff
@@ -62,7 +57,7 @@ def main():
     for obj in tStream:
         print obj
-        pass
+#        pass
```
Remove debug code.

mlyle replied:
If you want i can just delete minimal_connection.py and keep it outside of tree... it does nothing useful-- it is purely a scaffolding to develop the live telemetry interface with (and that i end up playing with as development continues)

solidgoldbomb replied:
Up to you.  It's nice to have an example user of the module but even examples should have debug/dead code removed from them.

solidgoldbomb commented on python/taulabs/telemetry.py:
```diff
@@ -177,28 +199,62 @@ def serviceConnection(self, timeout=None):
+        self.fd = fd
+
+    def _receive(self, finishTime):
+        """ Fetch available data from TCP socket """
```
This wouldn't have to be strictly a TCP socket now that you've abstracted out the FDTelemetry class.

solidgoldbomb commented on python/taulabs/telemetry.py:
```diff
@@ -212,36 +268,139 @@ def __select(self, finishTime):
-    def __receive(self, finishTime):
-        """ Fetch available data from TCP socket """
+    def _send(self, msg):
+        """ Send a string out the TCP socket """
```
This wouldn't have to be strictly a TCP socket now that you've abstracted out the FDTelemetry class.

mlyle replied:
Yup.  Next pass definitely has a new proper documentation pass with pydoc.

solidgoldbomb commented on python/taulabs/telemetry.py:
```diff
@@ -212,36 +268,139 @@ def __select(self, finishTime):
+    import os
+    return os.path.normpath(os.path.join(os.getcwd(), path))
+
+def GetUavoBasedOnArgs(desc=None):
```
Could (optionally) pass in a local argv here (with default of sys.argv) to allow callers first crack at parsing sys.argv or even generating their own args to the underlying telemetry decoder.

mlyle replied:
I thought about this.  argparse kinda assumes it's doing it all or doing none-- there's not an easy way if a parent called it to get the stuff that didn't match args (unless I missed it).   If they really care, they can instantiate FileTelemetry directly, I thought.

solidgoldbomb commented:
I really like the way you've collapsed the log file processing and telemetry streams into the same abstract class.  Nice.

solidgoldbomb commented:
Finished reviewing.  I would prefer to have something better than the ```readbyargs.py``` import side-effect mechanism before merging this.  Are you willing to take that on in this PR @mlyle?

mlyle commented:
@solidgoldbomb I'm willing to do whatever, but anything I do will be wordier than that.  My objective with the readbyargs stuff was to get something that would be as close as possible to the original ipython behavior with a minimal set of cruft needed to be appended to the end.

Let me know what you'd prefer instead.

mlyle commented:
(if you want, i can just remove it and replace it with its contents everywhere it's imported).

solidgoldbomb commented:
My vote would be what I wrote here: https://github.com/TauLabs/TauLabs/pull/1601#discussion_r30545185

mlyle commented:
That's like what I had before.  I can unwind later and submit that way.

mlyle commented:
@solidgoldbomb -- from IRC-- I don't really think they're making things worse.  It's a choice of two evils, between replicating 3-4 lines of boilerplate code in (potentially) a lot of places or by abusing a module import to have some side effects.  I originally had it close to the other way, and if that's preferred I'm fine with putting it back.. just need to get up the courage to do so :P

mlyle commented:
Done.

mlyle commented:
Yay, thanks for the merge. :)
